lib/devise-security/models/secure_validatable.rb

```diff
@@ -5,103 +5,218 @@
-      end
+        # validate the email address
+        # see https://github.com/devise-security/devise-security/blob/master/README.md#e-mail-validation
+        validates :email, email: email_validation if email_validation.present?
 
-      def current_equal_password_validation
-        return if new_record? || !will_save_change_to_encrypted_password? || password.blank?
-        dummy = self.class.new(encrypted_password: encrypted_password_was).tap do |user|
-          user.password_salt = password_salt_was if respond_to?(:password_salt)
+        # Check password complexity requirements
+        validate if: :password_required? do |record|
+          validates_with DeviseSecurity::PasswordComplexityValidator,
+                         { attributes: :password }.merge(record.password_complexity)
         end
-        self.errors.add(:password, :equal_to_current_password) if dummy.valid_password?(password)
-      end
 
-      protected
+        # Don't allow user to change the password back to the same one. Don't
+        # enforce this constraint if the stricter {PasswordArchivable} is in
+        # use.
+        validate :current_equal_password_validation,
+                 unless: ->(record) { record.devise_modules.include?(:password_archivable) }
+      end
 
-      # Checks whether a password is needed or not. For validations only.
-      # Passwords are always required if it's a new record, or if the password
-      # or confirmation are being set somewhere.
+      # Checks whether password validations are needed. Passwords are always
+      # required if it's a new record, or if the +password+ or
+      # +password_confirmation+ are set.
+      #
+      # @return [Boolean]
       def password_required?
-        !persisted? || !password.nil? || !password_confirmation.nil?
+        new_record? || password.present? || password_confirmation.present?
       end
 
+      # Requires an +email+ attribute to be defined. Override this method to
+      # return false to disable email validations.
+      #
+      # @return [Boolean]
       def email_required?
         true
       end
 
-      module ClassMethods
+      # Password complexity configurations for an instance of the base class,
+      # defaults to the class method. Override this for more fine-grained
+      # control of complexity configurations based on instance attributes.
+      #
+      # @return [Hash]
+      def password_complexity
+        self.class.password_complexity
+      end
+
+      # Password length requirements. Override this at the class or instance
+      # level to override global configurations.
+      #
+      # @return [Range]
+      def password_length
+        self.class.password_length
+      end
+
+      # Email validation options. Override this at the class or instance level
+      # to override global configurations.
+      #
+      # @return [Hash] email validation options
```

Comment:
```suggestion
      # @return [Boolean] whether email validation is enabled
```

Based on the default set up in `devise-security` it looks like this should be a Boolean:

https://github.com/devise-security/devise-security/blob/d11e44a7f496d88a66f6f82513a7b887ffa4ef01/lib/devise-security.rb#L29-L32

It also looks like we may not have added any tests for this method.